### PR TITLE
Implement Staking Period Calculator

### LINKED ISSUE #7580: Implement `EndOfStakingPeriodCalculator` from mono-service

### PR DESCRIPTION
**Pull request by mhess-swl**
Implementation of the staking period calculator that runs at the end of each staking period. 

**Note: this is not yet called from the handle workflow!**

Closes #7580

### CODE REVIEW

File: hedera-node/hedera-token-service-impl/src/main/java/com/hedera/node/app/service/token/impl/handlers/staking/EndOfStakingPeriodCalculator.java [hunk 1]
```diff
@@ -0,0 +1,412 @@
...
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import javax.inject.Inject;
+import javax.inject.Singleton;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+@Singleton
+public class EndOfStakingPeriodCalculator {
```
Comment (Neeharika-Sompalli) [inline hedera-node/hedera-token-service-impl/src/main/java/com/hedera/node/app/service/token/impl/handlers/staking/EndOfStakingPeriodCalculator.java:59]: Probably rename this class to `EndOfStakingPeriodUpdater` to not be confused to use wrong class in tests

File: hedera-node/hedera-token-service-impl/src/main/java/com/hedera/node/app/service/token/impl/handlers/staking/EndOfStakingPeriodCalculator.java [hunk 2]
```diff
@@ -0,0 +1,412 @@
...
+            @NonNull final HederaAccountNumbers accountNumbers, @NonNull final StakeInfoHelper stakeInfoHelper) {
+        this.accountNumbers = accountNumbers;
+        this.stakeInfoHelper = stakeInfoHelper;
+    }
+
+    /**
+     * Updates all (relevant) staking-related values for all nodes, as well as any network reward information,
+     * at the end of a staking period. This method must be invoked during handling of a transaction
+     *
+     * TODO: this method is still not called by the handle workflow
```
Comment (Neeharika-Sompalli) [inline hedera-node/hedera-token-service-impl/src/main/java/com/hedera/node/app/service/token/impl/handlers/staking/EndOfStakingPeriodCalculator.java:79]: We should probably create an issue and assign to hedera-base team or confirm if they already have issue for it

Comment (mhess-swl) [reply hedera-node/hedera-token-service-impl/src/main/java/com/hedera/node/app/service/token/impl/handlers/staking/EndOfStakingPeriodCalculator.java:79]: There is an issue for it, and I've coded it already. it's issue #7721

File: hedera-node/hedera-token-service-impl/src/main/java/com/hedera/node/app/service/token/impl/handlers/staking/EndOfStakingPeriodCalculator.java (hunk 1 above)
Comment (mhess-swl) [reply hedera-node/hedera-token-service-impl/src/main/java/com/hedera/node/app/service/token/impl/handlers/staking/EndOfStakingPeriodCalculator.java]: Done

File: hedera-node/hedera-token-service-impl/src/main/java/com/hedera/node/app/service/token/impl/handlers/staking/PeriodStakingUtils.java [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+import java.util.ArrayList;
+import java.util.List;
+import java.util.stream.IntStream;
+
+/**
+ * This class includes utility methods for dealing with staking period calculations (the updates to
+ * staking performed at the end of each staking period). That said, <b>these methods should NOT
+ * update any state</b> (e.g. by passing in a writable store)
+ */
+public final class PeriodStakingUtils {
```
Comment (Neeharika-Sompalli) [inline hedera-node/hedera-token-service-impl/src/main/java/com/hedera/node/app/service/token/impl/handlers/staking/PeriodStakingUtils.java:31]: Rename to `EndOfStakingperiodUtils` ?

File: hedera-node/hedera-token-service-impl/src/main/java/com/hedera/node/app/service/token/impl/handlers/staking/EndOfStakingPeriodUpdater.java [hunk 4]
```diff
@@ -0,0 +1,412 @@
...
+
+        // Calculate the updated stake and reward sum history for each node
+        long newTotalStakedStart = 0L;
+        long newTotalStakedRewardStart = 0L;
+        long maxStakeOfAllNodes = 0L;
+        final Map<Long, StakingNodeInfo> updatedNodeInfos = new HashMap<>();
+        for (final var nodeNum : nodeIds.stream().sorted().toList()) {
+            var currStakingInfo = stakingInfoStore.getForModify(nodeNum);
+
+            // The return value is the reward rate (tinybars-per-hbar-staked-to-reward) that will be paid to all
```
Comment (Neeharika-Sompalli) [inline hedera-node/hedera-token-service-impl/src/main/java/com/hedera/node/app/service/token/impl/handlers/staking/EndOfStakingPeriodUpdater.java:122]: Here return value is not just reward rate right ?

File: hedera-node/hedera-token-service-impl/src/main/java/com/hedera/node/app/service/token/impl/handlers/staking/EndOfStakingPeriodUpdater.java [hunk 5]
```diff
@@ -0,0 +1,412 @@
...
+            final var newStakeRewardStart = recomputedStake.stakeRewardStart();
+            final var nodePendingRewards = pendingRewardHbars * newPendingRewardRate;
+            log.info(
+                    "For node{}, the tb/hbar reward rate was {} for {} pending, with stake reward start {} -> {}",
+                    nodeNum,
+                    newPendingRewardRate,
+                    nodePendingRewards,
+                    oldStakeRewardStart,
+                    newStakeRewardStart);
+            stakeInfoHelper.increaseUnclaimedStakeRewards(nodeNum, nodePendingRewards, stakingInfoStore);
```
Comment (Neeharika-Sompalli) [inline hedera-node/hedera-token-service-impl/src/main/java/com/hedera/node/app/service/token/impl/handlers/staking/EndOfStakingPeriodUpdater.java:157]: I think these are pending rewards for next period and not `unclaimedRewards` . It seems the logic here is different

**Review by netopyr — APPROVED** [review]
LGTM in regards to `hedera-base`

File: hedera-node/hedera-token-service-impl/src/main/java/com/hedera/node/app/service/token/impl/handlers/staking/EndOfStakingPeriodUpdater.java (hunk 5 above)
Comment (mhess-swl) [reply hedera-node/hedera-token-service-impl/src/main/java/com/hedera/node/app/service/token/impl/handlers/staking/EndOfStakingPeriodUpdater.java:157]: Good catch, I used the wrong object for this! Will fix

File: hedera-node/hedera-token-service-impl/src/main/java/com/hedera/node/app/service/token/impl/handlers/staking/PeriodStakingUtils.java (hunk 3 above)
Comment (mhess-swl) [reply hedera-node/hedera-token-service-impl/src/main/java/com/hedera/node/app/service/token/impl/handlers/staking/PeriodStakingUtils.java]: Done

File: hedera-node/hedera-token-service-impl/src/main/java/com/hedera/node/app/service/token/impl/handlers/staking/EndOfStakingPeriodUpdater.java (hunk 4 above)
Comment (mhess-swl) [reply hedera-node/hedera-token-service-impl/src/main/java/com/hedera/node/app/service/token/impl/handlers/staking/EndOfStakingPeriodUpdater.java]: You're right, fixed this comment

File: hedera-node/hedera-token-service-impl/src/main/java/com/hedera/node/app/service/token/impl/handlers/staking/EndOfStakingPeriodUpdater.java (hunk 5 above)
Comment (mhess-swl) [reply hedera-node/hedera-token-service-impl/src/main/java/com/hedera/node/app/service/token/impl/handlers/staking/EndOfStakingPeriodUpdater.java]: Done

**Review by netopyr — APPROVED** [review]
LGTM in regards to `hedera-app`

**Review by Neeharika-Sompalli — APPROVED** [review]
LGTM ! Thanks @mhess-swl

**Pull request merged**